### PR TITLE
Add multiple language options to profiles

### PR DESCRIPTION
**Pull request by vawen**
Proposal for issue: Language Profiles #274 (https://github.com/Sonarr/Sonarr/issues/274)

This changes allows to have multiple languages on a profile working the same that qualities does.

Main changes:
- Added language to EpisodeFile. (Can be modified once set on Episode editor)
- Added UI interfaces to modify languages on EpisodeFiles (one by one and in batch)
- Modified blacklist and history to show language of the files.
- Removed Language enum, now it is a class.
- Added  a Language Profile (Set on Settings/Profile): Allows to select multiple download languages for a show. If more than one language is allowed Sonarr will select the one with more priority. Important, quality is always prefered over language.
- Rename of class: QualityUpgradableSpecification to UpgradableSpecification: Language is involved in upgrades now.
  Rename of variables: QualityUpgradableSpecification to upgradableSpecification (for consistency with the class renaming).

The changes are big, so feel free to contact me.

I hope you like it.

Edit: Updated to last commit after rebase and markus101 comments.

### CODE REVIEW

File: src/NzbDrone.Api/Profiles/ProfileModule.cs [hunk 1]
```diff
@@ -16,8 +17,9 @@ public ProfileModule(IProfileService profileService)
             _profileService = profileService;
             SharedValidator.RuleFor(c => c.Name).NotEmpty();
             SharedValidator.RuleFor(c => c.Cutoff).NotNull();
+            SharedValidator.RuleFor(c => c.CutoffLanguage).NotNull();
             SharedValidator.RuleFor(c => c.Items).MustHaveAllowedQuality();
-            SharedValidator.RuleFor(c => c.Language).ValidLanguage();
+			SharedValidator.RuleFor(c => c.Languages).MustHaveAllowedLanguage();
```
Comment (markus101) [inline src/NzbDrone.Api/Profiles/ProfileModule.cs]: Wrong indentation.

File: src/NzbDrone.Api/Profiles/LanguageValidator.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+using System.Linq;
+using FluentValidation;
+using FluentValidation.Validators;
+using NzbDrone.Api.Profiles;
+
+namespace NzbDrone.Core.Validation
+{
+    public static class LanguageValidation
+    {
+        public static IRuleBuilderOptions<T, IList<ProfileLanguageItemResource>> MustHaveAllowedLanguage<T>(this IRuleBuilder<T, IList<ProfileLanguageItemResource>> ruleBuilder)
```
Comment (markus101) [inline src/NzbDrone.Api/Profiles/LanguageValidator.cs]: `MustHaveAllowedLanguage` sounds a bit awkward to me, makes it sounds like we're testing a specification. Thoughts?

File: src/NzbDrone.Api/Profiles/ProfileSchemaModule.cs [hunk 3]
```diff
@@ -29,7 +30,11 @@ private List<ProfileResource> GetAll()
             var profile = new Profile();
             profile.Cutoff = Quality.Unknown;
             profile.Items = items;
-            profile.Language = Language.English;
+            profile.Languages = Language.All
```
Comment (markus101) [inline src/NzbDrone.Api/Profiles/ProfileSchemaModule.cs]: `.OrderByDescending(l => l.Name)` on the first line then align the rest on `.`

File: src/NzbDrone.Core.Test/Datastore/DatabaseRelationshipFixture.cs [hunk 4]
```diff
@@ -17,6 +18,7 @@ public void one_to_one()
         {
             var episodeFile = Builder<EpisodeFile>.CreateNew()
                            .With(c => c.Quality = new QualityModel())
```
Comment (markus101) [inline src/NzbDrone.Core.Test/Datastore/DatabaseRelationshipFixture.cs]: Align on `.`.

File: src/NzbDrone.Core/DecisionEngine/DownloadDecisionPriorizationService.cs [hunk 5]
```diff
@@ -30,15 +31,26 @@ public List<DownloadDecision> PrioritizeDecisions(List<DownloadDecision> decisio
...
-
-                                    return downloadDecisions
-                                        .OrderByDescending(c => c.RemoteEpisode.ParsedEpisodeInfo.Quality, new QualityModelComparer(series.Profile))
-                                        .ThenBy(c => c.RemoteEpisode.Episodes.Select(e => e.EpisodeNumber).MinOrDefault())
-                                        .ThenBy(c => PrioritizeDownloadProtocol(series, c.RemoteEpisode.Release.DownloadProtocol))
-                                        .ThenByDescending(c => c.RemoteEpisode.Episodes.Count)
-                                        .ThenBy(c => c.RemoteEpisode.Release.Size.Round(200.Megabytes()) / Math.Max(1, c.RemoteEpisode.Episodes.Count))
-                                        .ThenByDescending(c => TorrentInfo.GetSeeders(c.RemoteEpisode.Release))
-                                        .ThenBy(c => c.RemoteEpisode.Release.Age);
+                                    if (series.Profile.Value.LanguageOverQuality)
```
Comment (markus101) [inline src/NzbDrone.Core/DecisionEngine/DownloadDecisionPriorizationService.cs]: Too much duplication here, only the first lines are changing, but 6 other lines are being copied, since we're dealing with IEnumerables here nothing will be evaluated until the end, so we can order by the first two clauses (in the appropriate order) then add the other six.

File: src/NzbDrone.Core/DecisionEngine/QualityUpgradableSpecification.cs [hunk 6]
```diff
@@ -6,8 +7,8 @@ namespace NzbDrone.Core.DecisionEngine
 {
     public interface IQualityUpgradableSpecification
```
Comment (markus101) [inline src/NzbDrone.Core/DecisionEngine/QualityUpgradableSpecification.cs]: This needs a new name, its now combining quality and language, which might be required, if so it needs a new name, if we can come up with a way to separate these further then even better.

**Comment by markus101** [comment]
I still need to go through this some more, but I'm concerned that language and quality are so intertwined throughout this.

- Should we separate language and quality profiles? It adds some complexity to series, but keeps things more flexible, although language overriding quality becomes a conflict between two profiles. Maybe a top level profile that encompasses both (just throwing ideas out there)

- We need a way to separate the dependency between quality and language, I realize someone may want a quality downgrade to get a language upgrade (which we need to handle), but preferred words could also add to this complexity (assuming we let it override existing files, but not necessarily quality).

**Comment by thezoggy** [comment]
language can mean a few different things.. is the goal to grab releases in a foreign lang? is it to have your metadata in a different lang (by scraping from tvdb for a different lang than english - different tvdbids)..  

Lost in multiple languages:
http://thetvdb.com/index.php?string=Lost&searchseriesid=&tab=listseries&function=Search&alllang=1

File: src/NzbDrone.Api/Profiles/LanguageValidator.cs (hunk 2 above)
Comment (vawen) [inline src/NzbDrone.Api/Profiles/LanguageValidator.cs]: It is pretty much the same that with quality. Maybe change both to MustHaveAnAllowed<Quality/Language>?

File: src/NzbDrone.Core/DecisionEngine/QualityUpgradableSpecification.cs (hunk 6 above)
Comment (vawen) [inline src/NzbDrone.Core/DecisionEngine/QualityUpgradableSpecification.cs]: I see the separation a bit hard. As I see it you have 3 scenarios:
1. You want best quality and once a language is donwloaded you want to stick with it (current situation).
2. You want best quality and doesn't care about language given a list (new situation, you can upgrade language if quality doesn't get downgrade).
3. You want best language possible, even if that mean you downgrade quality (new scenario)

I think the three are linked to the "quality" of the file downloaded, since the language is also a quality measure, at least IMO.

**Comment by vawen** [comment]
thezoggy: Yes, the goal is to download foreign lang files.

File: src/NzbDrone.Core/DecisionEngine/QualityUpgradableSpecification.cs (hunk 6 above)
Comment (vawen) [inline src/NzbDrone.Core/DecisionEngine/QualityUpgradableSpecification.cs]: Completly agree. I think IUpgradableSpecification is more inline with the work that class is doing atm.

**Comment by vawen** [comment]
Did a rebase to current state.

File: src/NzbDrone.Api.Test/MappingTests/ResourceMappingFixture.cs [hunk 7]
```diff
@@ -109,14 +113,29 @@ public void should_map_profile()
...
+        [Test]
+        public void should_map_language_profile()
+        {
+            var profileResource = new LanguageProfileResource
+            {
+                Languages = new List<ProfileLanguageItemResource> { new ProfileLanguageItemResource { Language = Language.English, Allowed = true } },
+                Cutoff = Language.English
+            };
+
+
```
Comment (markus101) [inline src/NzbDrone.Api.Test/MappingTests/ResourceMappingFixture.cs]: Extra whitespace should be cleaned up.

File: src/NzbDrone.Api.Test/MappingTests/ResourceMappingFixture.cs [hunk 8]
```diff
@@ -109,14 +113,29 @@ public void should_map_profile()
             var profileResource = new ProfileResource
                 {
                     Cutoff = Quality.WEBDL1080p,
-                    Items = new List<ProfileQualityItemResource> { new ProfileQualityItemResource { Quality = Quality.WEBDL1080p, Allowed = true } }
+                    Items = new List<ProfileQualityItemResource> { new ProfileQualityItemResource { Quality = Quality.WEBDL1080p, Allowed = true } },
```
Comment (markus101) [inline src/NzbDrone.Api.Test/MappingTests/ResourceMappingFixture.cs]: Can drop the comma again.

File: src/NzbDrone.Api/LanguageProfiles/LanguageProfileResource.cs [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        public List<ProfileLanguageItemResource> Languages { get; set; }
+    }
+
+    public class ProfileLanguageItemResource : RestResource
+    {
+        public Language Language { get; set; }
+        public bool Allowed { get; set; }
+    }
+
+}
```
Comment (markus101) [inline src/NzbDrone.Api/LanguageProfiles/LanguageProfileResource.cs]: New line at end of file.

File: src/NzbDrone.Api/LanguageProfiles/Languages/LanguageModule.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NzbDrone.Core.Parser;
+using NzbDrone.Core.Languages;
+
+namespace NzbDrone.Api.LanguageProfiles.Languages
+{
+    //TODO: Deprecated with LanguageProfileSchema it is not used now
```
Comment (markus101) [inline src/NzbDrone.Api/LanguageProfiles/Languages/LanguageModule.cs]: What do we need to do here?

File: src/NzbDrone.Api/LanguageProfiles/Languages/LanguageResource.cs [hunk 11]
```diff
@@ -2,8 +2,9 @@
 using Newtonsoft.Json;
 using NzbDrone.Api.REST;
 
-namespace NzbDrone.Api.Profiles.Languages
+namespace NzbDrone.Api.LanguageProfiles.Languages
 {
+    //TODO: No longer used with LanguageProfileSchema
```
Comment (markus101) [inline src/NzbDrone.Api/LanguageProfiles/Languages/LanguageResource.cs]: Same here, what do we need to do?

File: src/NzbDrone.Api/LanguageProfiles/languageProfileModule.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+using FluentValidation;
+using NzbDrone.Api.Mapping;
+using NzbDrone.Core.Profiles;
+using NzbDrone.Core.Qualities;
+using NzbDrone.Core.Validation;
+using NzbDrone.Core.Languages;
+
+namespace NzbDrone.Api.LanguageProfiles
+{
+    public class LanguageProfileModule : NzbDroneRestModule<LanguageProfileResource>
```
Comment (markus101) [inline src/NzbDrone.Api/LanguageProfiles/languageProfileModule.cs]: File name needs to be capitalized.

File: src/NzbDrone.Api/NzbDrone.Api.csproj [hunk 13]
```diff
@@ -194,11 +198,11 @@
...
     <Compile Include="NzbDroneRestModuleWithSignalR.cs" />
     <Compile Include="PagingResource.cs" />
-    <Compile Include="Profiles\Languages\LanguageModule.cs" />
-    <Compile Include="Profiles\Languages\LanguageResource.cs" />
+    <Compile Include="LanguageProfiles\Languages\LanguageModule.cs" />
+    <Compile Include="LanguageProfiles\Languages\LanguageResource.cs" />
     <Compile Include="Profiles\LegacyProfileModule.cs" />
-    <Compile Include="Profiles\ProfileModule.cs" />
-    <Compile Include="Profiles\ProfileResource.cs" />
+    <Compile Include="LanguageProfiles\languageProfileModule.cs" />
```
Comment (markus101) [inline src/NzbDrone.Api/NzbDrone.Api.csproj]: I think `Profiles\Quality` and `Profiles\Language` makes sense to locate these two somewhat related items.

File: src/NzbDrone.Api/Profiles/ProfileSchemaModule.cs [hunk 14]
```diff
@@ -4,6 +4,7 @@
 using NzbDrone.Core.Parser;
 using NzbDrone.Core.Profiles;
 using NzbDrone.Core.Qualities;
+using NzbDrone.Core.Languages;
```
Comment (markus101) [inline src/NzbDrone.Api/Profiles/ProfileSchemaModule.cs]: Not required.

File: src/NzbDrone.Core.Test/Datastore/MarrDataLazyLoadingFixture.cs [hunk 15]
```diff
@@ -100,8 +111,27 @@ public void should_explicit_load_profile_if_joined()
...
         }
 
+        [Test]
+        public void should_explicit_load_languageprofile_if_joined()
+        {
+            var db = Mocker.Resolve<IDatabase>();
+            var DataMapper = db.GetDataMapper();
+
+            var episodes = DataMapper.Query<Episode>()
+                .Join<Episode, Series>(Marr.Data.QGen.JoinType.Inner, v => v.Series, (l, r) => l.SeriesId == r.Id)
```
Comment (markus101) [inline src/NzbDrone.Core.Test/Datastore/MarrDataLazyLoadingFixture.cs]: Align on `.`

File: src/NzbDrone.Core.Test/Datastore/Migration/075_force_lib_updateFixture.cs [hunk 16]
```diff
@@ -57,6 +57,14 @@ public void should_reset_series_last_sync_time()
         {
             WithTestDb(c =>
             {
+                c.Insert.IntoTable("Profiles").Row(new
```
Comment (markus101) [inline src/NzbDrone.Core.Test/Datastore/Migration/075_force_lib_updateFixture.cs]: Why do these old tests need to be updated? Changing a migration test is a bit of a concern because we need to be very sure nothing breaks for people running old DB versions.

File: src/NzbDrone.Core.Test/DecisionEngineTests/HistorySpecificationFixture.cs [hunk 17]
```diff
@@ -46,30 +46,41 @@ public void Setup()
...
+                             })
                          .Build();
 
             _parseResultMulti = new RemoteEpisode
             {
                 Series = _fakeSeries,
-                ParsedEpisodeInfo = new ParsedEpisodeInfo { Quality = new QualityModel(Quality.DVD, new Revision(version: 2)) },
-                Episodes = doubleEpisodeList
+                ParsedEpisodeInfo = new ParsedEpisodeInfo { Quality = new QualityModel(Quality.DVD, new Revision(version: 2)), Language = Language.English },
+                Episodes = doubleEpisodeList,
```
Comment (markus101) [inline src/NzbDrone.Core.Test/DecisionEngineTests/HistorySpecificationFixture.cs]: Whitespace gone wild.

File: src/NzbDrone.Core.Test/DecisionEngineTests/LanguageSpecificationFixture.cs [hunk 18]
```diff
@@ -40,6 +47,17 @@ private void WithEnglishRelease()
...
+        {
+            _remoteEpisode.ParsedEpisodeInfo.Language = Language.Spanish;
+        }
+
+        private void WithFrenchRelease()
+        {
+            _remoteEpisode.ParsedEpisodeInfo.Language = Language.French;
+        }
+
+
```
Comment (markus101) [inline src/NzbDrone.Core.Test/DecisionEngineTests/LanguageSpecificationFixture.cs]: Whitespace.

File: src/NzbDrone.Core.Test/ParserTests/LanguageParserFixture.cs [hunk 19]
```diff
@@ -9,47 +10,47 @@ namespace NzbDrone.Core.Test.ParserTests
...
-        [TestCase("Revolution S01E03 No Quarter 2012 WEB-DL 720p Nordic-philipo mkv", Language.Norwegian)]
-        [TestCase("Extant.S01E01.VOSTFR.HDTV.x264-RiDERS", Language.French)]
-        [TestCase("Constantine.2014.S01E01.WEBRiP.H264.AAC.5.1-NL.SUBS", Language.Dutch)]
-        [TestCase("Elementary - S02E16 - Kampfhaehne - mkv - by Videomann", Language.German)]
-        [TestCase("Two.Greedy.Italians.S01E01.The.Family.720p.HDTV.x264-FTP", Language.English)]
-        [TestCase("Castle.2009.S01E14.HDTV.XviD.HUNDUB-LOL", Language.Hungarian)]
-        [TestCase("Castle.2009.S01E14.HDTV.XviD.ENG.HUN-LOL", Language.Hungarian)]
-        [TestCase("Castle.2009.S01E14.HDTV.XviD.HUN-LOL", Language.Hungarian)]
-        public void should_parse_language(string postTitle, Language language)
+        [TestCase("Castle.2009.S01E14.English.HDTV.XviD-LOL", 1)]
```
Comment (markus101) [inline src/NzbDrone.Core.Test/ParserTests/LanguageParserFixture.cs]: I think we should do something similar to the quality tests where we have test methods that test specific languages, as-is these are incredibly hard to follow because I need to have a mental map for language and ID.

File: src/NzbDrone.Core.Test/TvTests/EpisodeRepositoryTests/EpisodesWhereCutoffUnmetFixture.cs [hunk 20]
```diff
@@ -37,12 +39,25 @@ public void Setup()
                 }
             };
 
+            var langProfile = new LanguageProfile
+            {
+                Id = 1,
+                Languages = new List<ProfileLanguageItem>
```
Comment (markus101) [inline src/NzbDrone.Core.Test/TvTests/EpisodeRepositoryTests/EpisodesWhereCutoffUnmetFixture.cs]: There seems to be a lot of the same object creation, we get a big chuck of code cut out by creating a default profile we can use for testing.

File: src/NzbDrone.Core.Test/HistoryTests/HistoryServiceFixture.cs [hunk 21]
```diff
@@ -21,46 +22,93 @@ public class HistoryServiceFixture : CoreTest<HistoryService>
...
+                });
 
-            var quality = Subject.GetBestQualityInHistory(_profile, 2);
+            var quality = Subject.GetBestInHistory(_profile, _languageProfile, 2).Quality;
 
             quality.Should().Be(new QualityModel(Quality.Bluray1080p));
         }
 
         [Test]
+        public void should_return_best_quality_with_language()
```
Comment (markus101) [inline src/NzbDrone.Core.Test/HistoryTests/HistoryServiceFixture.cs]: I recently removed one use of the best quality in history method, I'll double check but it might be able to get killed off completely.

File: src/NzbDrone.Core.Test/MediaFiles/EpisodeImport/ImportDecisionMakerFixture.cs [hunk 22]
```diff
@@ -55,6 +56,16 @@ public void Setup()
 
             _series = Builder<Series>.CreateNew()
                                      .With(e => e.Profile = new Profile { Items = Qualities.QualityFixture.GetDefaultQualities() })
+                                     .With(l => l.LanguageProfile =
+                                         new LanguageProfile
+                                         {
+                                             Languages = new List<ProfileLanguageItem>
```
Comment (markus101) [inline src/NzbDrone.Core.Test/MediaFiles/EpisodeImport/ImportDecisionMakerFixture.cs]: Same here, we're creating a new Language profile multiple times, but for quality profile we just call:

`.With(e => e.Profile = new Profile { Items = Qualities.QualityFixture.GetDefaultQualities() })`

File: src/NzbDrone.Core.Test/MediaFiles/EpisodeImport/Specifications/UpgradeSpecificationFixture.cs [hunk 23]
```diff
@@ -22,15 +24,31 @@ public class UpgradeSpecificationFixture : CoreTest<UpgradeSpecification>
...
+                                        })
+                                     .With(l => l.LanguageProfile = new LanguageProfile
+                                        {
+                                            Languages = _languages,
+                                            Cutoff = Language.Spanish,
+                                        })
                                      .Build();
 
-            _localEpisode = new LocalEpisode
+             _localEpisode = new LocalEpisode
```
Comment (markus101) [inline src/NzbDrone.Core.Test/MediaFiles/EpisodeImport/Specifications/UpgradeSpecificationFixture.cs]: Alignment.

File: src/NzbDrone.Core/Datastore/Migration/095_add_language_to_episodeFiles.cs [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+using System.IO;
+using FluentMigrator;
+using NzbDrone.Core.Datastore.Migration.Framework;
+using NzbDrone.Core.Datastore.Converters;
+using NzbDrone.Core.Languages;
+using System;
+
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(95)]
```
Comment (markus101) [inline src/NzbDrone.Core/Datastore/Migration/095_add_language_to_episodeFiles.cs]: Looks like one was renamed and the other wasn't?

File: src/NzbDrone.Core/Datastore/Migration/120_blacklist_language.cs [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+using System.Linq;
+using FluentMigrator;
+using NzbDrone.Common.Extensions;
+using NzbDrone.Common.Serializer;
+using NzbDrone.Core.Datastore.Migration.Framework;
+
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(120)]
+    public class blacklist_language : NzbDroneMigrationBase
```
Comment (markus101) [inline src/NzbDrone.Core/Datastore/Migration/120_blacklist_language.cs]: Why 4 separate migrations all numbered somewhat randomly? I think we could get away with 2, one that adds language profiles and one that add language it to the appropriate tables.

File: src/NzbDrone.Core/DecisionEngine/QualityUpgradableSpecification.cs [hunk 26]
```diff
@@ -1,74 +0,0 @@
-using NLog;
-using NzbDrone.Core.Profiles;
-using NzbDrone.Core.Qualities;
-
-namespace NzbDrone.Core.DecisionEngine
-{
-    public interface IQualityUpgradableSpecification
```
Comment (markus101) [inline src/NzbDrone.Core/DecisionEngine/QualityUpgradableSpecification.cs]: Wouldn't think still be used by UpgradableSpec?

File: src/NzbDrone.Core/DecisionEngine/UpgradableSpecification.cs [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+        public UpgradableSpecification(Logger logger)
+        {
+            _logger = logger;
+        }
+
+        private bool IsLanguageUpgradable(LanguageProfile profile, Language currentLanguage, Language newLanguage = null) 
+        {
+            if (newLanguage != null)
+            {
+                int compare = new LanguageComparer(profile).Compare(newLanguage, currentLanguage);
```
Comment (markus101) [inline src/NzbDrone.Core/DecisionEngine/UpgradableSpecification.cs]: `var` instead of `int`

File: src/NzbDrone.Core/DecisionEngine/UpgradableSpecification.cs [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+                if (compare <= 0)
+                    return false;
```
Comment (markus101) [inline src/NzbDrone.Core/DecisionEngine/UpgradableSpecification.cs]: Always use `{}`s around if blocks.

File: src/NzbDrone.Core/DecisionEngine/UpgradableSpecification.cs [hunk 29]
```diff
@@ -0,0 +1,104 @@
...
+                return false;
+            }
+
+            return true;
+        }
+
+        public bool CutoffNotMet(Profile profile, LanguageProfile languageProfile, QualityModel currentQuality, Language currentLanguage, QualityModel newQuality = null)
+        {
+            int languageCompare = new LanguageComparer(languageProfile).Compare(currentLanguage, languageProfile.Cutoff);
+            int compare = new QualityModelComparer(profile).Compare(currentQuality.Quality, profile.Cutoff);
```
Comment (markus101) [inline src/NzbDrone.Core/DecisionEngine/UpgradableSpecification.cs]: `qualityCompare`

**Comment by markus101** [comment]
Overall this is pretty good, I think there is a lot of room for improvement in how the tests are written to make them more clear and less duplication in them.

I'll probably need to review the UI changes again because I didn't look too closely at them.

For the DB migrations to save you pain when we add a new migration I think we should add a couple placeholder files so we see them in source explorer, but they don't execute anything and we can use them to make sure we don't keep nuking the numbering that you used because it was available when you created it.

File: src/NzbDrone.Core.Test/Datastore/Migration/075_force_lib_updateFixture.cs (hunk 16 above)
Comment (vawen) [inline src/NzbDrone.Core.Test/Datastore/Migration/075_force_lib_updateFixture.cs]: For testing. Or it will fail on import db tests.

File: src/NzbDrone.Core/DecisionEngine/QualityUpgradableSpecification.cs (hunk 26 above)
Comment (vawen) [inline src/NzbDrone.Core/DecisionEngine/QualityUpgradableSpecification.cs]: Modified the name to IUpgradableSpecification since language is also involved in the method now.

**Comment by JigSawFr** [comment]
Nice PR, What's next step before merging @markus101 ?

**Comment by markus101** [comment]
At the very least this needs review from @Taloth or @kayone plus a decision on whether this waits for v3 as we're changing the API by moving the language out of profile.

**Comment by JigSawFr** [comment]
@Taloth @kayone @markus101 
Any news ?

**Comment by foux** [comment]
I would also like to have news on this. This is a feature I'd like a lot, and the PR has been opened for quite some time now ;)
@JigSawFr perhaps you could resolve the current conflicts so that it can be reviewed by the person involved?

**Comment by JigSawFr** [comment]
Waiting for this long awaited feature :dart:

**Comment by vawen** [comment]
@foux I have been busy with work. Due to migration number 101, the migration I have will cause problems now. @markus101 I think it will be easier if I move the language profile migration to 102, rather than modify migration 101. It is ok?

**Comment by markus101** [comment]
Sorry @vawen yes, lets make it 102, definitely can't alter migration 101 as its already deployed. Sorry about that.

As for implementation, this is very likely going to be part of v3 due to the API changes.

**Comment by vawen** [comment]
Updated to current develop branch. I used migrations 102 and 104 Markus101, let me know if prefer others.

**Comment by DiseaseNO** [comment]
Any new on this?

**Comment by foXaCe** [comment]
hello, perhaps this approach is good

https://github.com/Radarr/Radarr/commit/8e452f8b6d3977559c5eebcd6e08bfc08835fa44

**Comment by markus101** [comment]
This has landed in the v3 work. There is no ETA for when v3 will be available.
